### ISSUE TITLE
Consider cleaning up `spanner_internal::PartialResultSetReader` interface

### ISSUE THREAD

**Issue by dbolduc** [googleapis/google-cloud-cpp#11888]
This cleanup is leftover from the refactor in #11857.

The `PartialResultSetReader` interface used to directly wrap a gRPC stream. Now it is implemented by a `google::cloud::StreamingReadRpc` (which calls `Finish()` on the underlying gRPC stream for us). I think we want an interface that matches that of the `StreamingReadRpc`.

```cc
class PartialResultSetReader {
 public:
  void Cancel();
  absl::variant<Status, PartialResultSet> Read(absl::optional<std::string> resume) = 0;
  // No more `Finish()`
};
```

Note that we just remember the status from the final `StreamingReadRpc<>::Read()` and return that.

https://github.com/googleapis/google-cloud-cpp/blob/1145a8ccff7f8ce7fcebc0256efe6b3093e8e21a/google/cloud/spanner/internal/connection_impl.cc#L75-L92

---

Then there are the places in the code which interact with this interface. They manually call `PartialResultSetReader::Finish()`. I think we can do away with `kEndOfStream` altogether.

https://github.com/googleapis/google-cloud-cpp/blob/1145a8ccff7f8ce7fcebc0256efe6b3093e8e21a/google/cloud/spanner/internal/partial_result_set_source.cc#L127-L136

---

Basically, the code could be simplified.

**Comment by devbww** [comment]
Probably want/need to do this cleanup when looking at #11890.
